File: python/networking/zip_lookup.py
"""Zip / Postal Code Lookup — resolve postal codes to location data."""

from dataclasses import dataclass, field
from typing import Optional
import math


@dataclass
class PostalResult:
    """Location data for a postal code."""
    postal_code: str
    country_code: str = ""
    country_name: str = ""
    state: str = ""
    state_abbr: str = ""
    city: str = ""
    latitude: float = 0.0
    longitude: float = 0.0
    timezone: str = ""

    def summary(self) -> str:
        parts = [p for p in [self.city, self.state_abbr or self.state, self.country_code] if p]
        loc = ", ".join(parts)
        return f"{self.postal_code} -> {loc} [{self.latitude:.4f}, {self.longitude:.4f}]"


@dataclass
class PostalEntry:
    """Database entry for a postal code."""
    postal_code: str
    country_code: str
    country_name: str
    state: str
    state_abbr: str
    city: str
    latitude: float
    longitude: float
    timezone: str = ""


# Built-in database: a selection of US and international postal codes.
BUILTIN_ENTRIES: list[PostalEntry] = [
    # Florida
    PostalEntry("32099", "US", "United States", "Florida", "FL", "Jacksonville", 30.3322, -81.6557, "America/New_York"),
    PostalEntry("32202", "US", "United States", "Florida", "FL", "Jacksonville", 30.3280, -81.6550, "America/New_York"),
    PostalEntry("32244", "US", "United States", "Florida", "FL", "Jacksonville", 30.2266, -81.7360, "America/New_York"),
    PostalEntry("32084", "US", "United States", "Florida", "FL", "St. Augustine", 29.8947, -81.3145, "America/New_York"),
    PostalEntry("33101", "US", "United States", "Florida", "FL", "Miami", 25.7617, -80.1918, "America/New_York"),
    PostalEntry("32801", "US", "United States", "Florida", "FL", "Orlando", 28.5383, -81.3792, "America/New_York"),
    # Other US
    PostalEntry("10001", "US", "United States", "New York", "NY", "New York", 40.7484, -73.9967, "America/New_York"),
    PostalEntry("10013", "US", "United States", "New York", "NY", "New York", 40.7209, -74.0048, "America/New_York"),
    PostalEntry("90210", "US", "United States", "California", "CA", "Beverly Hills", 34.0901, -118.4065, "America/Los_Angeles"),
    PostalEntry("94105", "US", "United States", "California", "CA", "San Francisco", 37.7893, -122.3932, "America/Los_Angeles"),
    PostalEntry("02101", "US", "United States", "Massachusetts", "MA", "Boston", 42.3601, -71.0589, "America/New_York"),
    PostalEntry("60601", "US", "United States", "Illinois", "IL", "Chicago", 41.8819, -87.6278, "America/Chicago"),
    PostalEntry("78201", "US", "United States", "Texas", "TX", "San Antonio", 29.4685, -98.5254, "America/Chicago"),
    PostalEntry("98101", "US", "United States", "Washington", "WA", "Seattle", 47.6062, -122.3321, "America/Los_Angeles"),
    # International
    PostalEntry("SW1A 1AA", "GB", "United Kingdom", "England", "", "London", 51.5014, -0.1419, "Europe/London"),
    PostalEntry("EC1A 1BB", "GB", "United Kingdom", "England", "", "London", 51.5203, -0.0985, "Europe/London"),
    PostalEntry("75001", "FR", "France", "Île-de-France", "", "Paris", 48.8606, 2.3376, "Europe/Paris"),
    PostalEntry("10115", "DE", "Germany", "Berlin", "", "Berlin", 52.5326, 13.3884, "Europe/Berlin"),
    PostalEntry("100-0001", "JP", "Japan", "Tokyo", "", "Chiyoda", 35.6895, 139.6917, "Asia/Tokyo"),
    PostalEntry("2000", "AU", "Australia", "New South Wales", "NSW", "Sydney", -33.8688, 151.2093, "Australia/Sydney"),
]


class PostalDatabase:
    """Postal code database with lookup, search, and distance operations."""

    def __init__(self, entries: Optional[list[PostalEntry]] = None) -> None:
        self._by_code: dict[str, PostalEntry] = {}
        self._by_city: dict[str, list[PostalEntry]] = {}
        self._by_state: dict[str, list[PostalEntry]] = {}
        for e in (entries or BUILTIN_ENTRIES):
            self.add_entry(e)

    def add_entry(self, e: PostalEntry) -> None:
        self._by_code[e.postal_code.upper().replace(" ", "")] = e
        city_key = e.city.lower()
        self._by_city.setdefault(city_key, []).append(e)
        state_key = (e.state_abbr or e.state).lower()
        self._by_state.setdefault(state_key, []).append(e)

    def lookup(self, postal_code: str) -> Optional[PostalResult]:
        """Look up a postal code."""
        key = postal_code.upper().replace(" ", "")
        e = self._by_code.get(key)
        if e is None:
            return None
        return PostalResult(
            postal_code=e.postal_code, country_code=e.country_code,
            country_name=e.country_name, state=e.state, state_abbr=e.state_abbr,
            city=e.city, latitude=e.latitude, longitude=e.longitude, timezone=e.timezone,
        )

    def search_by_city(self, city: str) -> list[PostalResult]:
        """Find all postal codes for a city."""
        entries = self._by_city.get(city.lower(), [])
        return [PostalResult(
            postal_code=e.postal_code, country_code=e.country_code,
            country_name=e.country_name, state=e.state, state_abbr=e.state_abbr,
            city=e.city, latitude=e.latitude, longitude=e.longitude, timezone=e.timezone,
        ) for e in entries]

    def search_by_state(self, state: str) -> list[PostalResult]:
        """Find all postal codes in a state."""
        entries = self._by_state.get(state.lower(), [])
        return [PostalResult(
            postal_code=e.postal_code, country_code=e.country_code,
            country_name=e.country_name, state=e.state, state_abbr=e.state_abbr,
            city=e.city, latitude=e.latitude, longitude=e.longitude, timezone=e.timezone,
        ) for e in entries]

    def bulk_lookup(self, codes: list[str]) -> list[Optional[PostalResult]]:
        return [self.lookup(c) for c in codes]

    def distance_between(self, code_a: str, code_b: str) -> Optional[float]:
        """Return distance in km between two postal codes."""
        a = self.lookup(code_a)
        b = self.lookup(code_b)
        if a is None or b is None:
            return None
        return haversine_km(a.latitude, a.longitude, b.latitude, b.longitude)

    def codes_within_radius(self, center_code: str, radius_km: float) -> list[tuple[PostalResult, float]]:
        """Find all postal codes within radius_km of center_code."""
        center = self.lookup(center_code)
        if center is None:
            return []
        results = []
        for e in self._by_code.values():
            d = haversine_km(center.latitude, center.longitude, e.latitude, e.longitude)
            if d <= radius_km and e.postal_code != center.postal_code:
                results.append((PostalResult(
                    postal_code=e.postal_code, country_code=e.country_code,
                    country_name=e.country_name, state=e.state, state_abbr=e.state_abbr,
                    city=e.city, latitude=e.latitude, longitude=e.longitude, timezone=e.timezone,
                ), d))
        results.sort(key=lambda x: x[1])
        return results

    def state_breakdown(self, codes: list[str]) -> dict[str, int]:
        """Count postal codes per state."""
        counts: dict[str, int] = {}
        for c in codes:
            r = self.lookup(c)
            key = (r.state_abbr or r.state) if r else "unknown"
            counts[key] = counts.get(key, 0) + 1
        return counts


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    R = 6371.0
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (math.sin(dlat / 2) ** 2 +
         math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) *
         math.sin(dlon / 2) ** 2)
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


# --- demo ---
if __name__ == "__main__":
    db = PostalDatabase()
    for code in ["32202", "10001", "90210", "SW1A 1AA", "75001"]:
        r = db.lookup(code)
        print(r.summary() if r else f"{code} -> not found")
    print()
    d = db.distance_between("32202", "32084")
    print(f"Jacksonville to St. Augustine: {d:.1f} km")
    print()
    nearby = db.codes_within_radius("32202", 500)
    print(f"Codes within 500km of Jacksonville: {len(nearby)}")
    for r, dist in nearby:
        print(f"  {r.postal_code} {r.city}: {dist:.0f} km")

File: go/networking/zip_lookup.go
// Zip / Postal Code Lookup — resolve postal codes to location data.
package networking

import (
	"fmt"
	"math"
	"sort"
	"strings"
)

// PostalResult holds location data for a postal code.
type PostalResult struct {
	PostalCode  string
	CountryCode string
	CountryName string
	State       string
	StateAbbr   string
	City        string
	Latitude    float64
	Longitude   float64
	Timezone    string
}

// Summary returns a one-line description.
func (r PostalResult) PostalSummary() string {
	st := r.StateAbbr
	if st == "" {
		st = r.State
	}
	var parts []string
	for _, s := range []string{r.City, st, r.CountryCode} {
		if s != "" {
			parts = append(parts, s)
		}
	}
	return fmt.Sprintf("%s -> %s [%.4f, %.4f]", r.PostalCode, strings.Join(parts, ", "), r.Latitude, r.Longitude)
}

type postalEntry struct {
	PostalCode  string
	CountryCode string
	CountryName string
	State       string
	StateAbbr   string
	City        string
	Latitude    float64
	Longitude   float64
	Timezone    string
}

func (e *postalEntry) toResult() PostalResult {
	return PostalResult{e.PostalCode, e.CountryCode, e.CountryName, e.State, e.StateAbbr, e.City, e.Latitude, e.Longitude, e.Timezone}
}

func postalBuiltins() []postalEntry {
	e := func(code, cc, cn, st, sa, city string, lat, lon float64, tz string) postalEntry {
		return postalEntry{code, cc, cn, st, sa, city, lat, lon, tz}
	}
	return []postalEntry{
		e("32099", "US", "United States", "Florida", "FL", "Jacksonville", 30.3322, -81.6557, "America/New_York"),
		e("32202", "US", "United States", "Florida", "FL", "Jacksonville", 30.3280, -81.6550, "America/New_York"),
		e("32244", "US", "United States", "Florida", "FL", "Jacksonville", 30.2266, -81.7360, "America/New_York"),
		e("32084", "US", "United States", "Florida", "FL", "St. Augustine", 29.8947, -81.3145, "America/New_York"),
		e("33101", "US", "United States", "Florida", "FL", "Miami", 25.7617, -80.1918, "America/New_York"),
		e("10001", "US", "United States", "New York", "NY", "New York", 40.7484, -73.9967, "America/New_York"),
		e("90210", "US", "United States", "California", "CA", "Beverly Hills", 34.0901, -118.4065, "America/Los_Angeles"),
		e("94105", "US", "United States", "California", "CA", "San Francisco", 37.7893, -122.3932, "America/Los_Angeles"),
		e("SW1A1AA", "GB", "United Kingdom", "England", "", "London", 51.5014, -0.1419, "Europe/London"),
		e("75001", "FR", "France", "Île-de-France", "", "Paris", 48.8606, 2.3376, "Europe/Paris"),
		e("2000", "AU", "Australia", "New South Wales", "NSW", "Sydney", -33.8688, 151.2093, "Australia/Sydney"),
	}
}

func normalizePostalCode(code string) string {
	return strings.ToUpper(strings.ReplaceAll(code, " ", ""))
}

// PostalDatabase maps postal codes to locations.
type PostalDatabase struct {
	byCode  map[string]*postalEntry
	entries []postalEntry
}

// NewPostalDatabase creates a database with built-in entries.
func NewPostalDatabase() *PostalDatabase {
	entries := postalBuiltins()
	byCode := make(map[string]*postalEntry)
	for i := range entries {
		byCode[normalizePostalCode(entries[i].PostalCode)] = &entries[i]
	}
	return &PostalDatabase{byCode: byCode, entries: entries}
}

// PostalLookup resolves a postal code.
func (db *PostalDatabase) PostalLookup(code string) *PostalResult {
	e, ok := db.byCode[normalizePostalCode(code)]
	if !ok {
		return nil
	}
	r := e.toResult()
	return &r
}

// SearchByCity finds all codes for a city.
func (db *PostalDatabase) SearchByCity(city string) []PostalResult {
	cityLower := strings.ToLower(city)
	var results []PostalResult
	for _, e := range db.entries {
		if strings.ToLower(e.City) == cityLower {
			results = append(results, e.toResult())
		}
	}
	return results
}

// PostalDistanceBetween returns km between two postal codes.
func (db *PostalDatabase) PostalDistanceBetween(a, b string) (float64, bool) {
	ra := db.PostalLookup(a)
	rb := db.PostalLookup(b)
	if ra == nil || rb == nil {
		return 0, false
	}
	return postalHaversine(ra.Latitude, ra.Longitude, rb.Latitude, rb.Longitude), true
}

// CodesWithinRadius finds codes within radius_km of center.
func (db *PostalDatabase) CodesWithinRadius(center string, radiusKm float64) []struct {
	Result   PostalResult
	Distance float64
} {
	c := db.PostalLookup(center)
	if c == nil {
		return nil
	}
	centerNorm := normalizePostalCode(center)
	type rd struct {
		Result   PostalResult
		Distance float64
	}
	var results []rd
	for _, e := range db.entries {
		if normalizePostalCode(e.PostalCode) == centerNorm {
			continue
		}
		d := postalHaversine(c.Latitude, c.Longitude, e.Latitude, e.Longitude)
		if d <= radiusKm {
			results = append(results, rd{e.toResult(), d})
		}
	}
	sort.Slice(results, func(i, j int) bool { return results[i].Distance < results[j].Distance })
	out := make([]struct {
		Result   PostalResult
		Distance float64
	}, len(results))
	for i, r := range results {
		out[i] = struct {
			Result   PostalResult
			Distance float64
		}{r.Result, r.Distance}
	}
	return out
}

func postalHaversine(lat1, lon1, lat2, lon2 float64) float64 {
	const R = 6371.0
	dLat := (lat2 - lat1) * math.Pi / 180
	dLon := (lon2 - lon1) * math.Pi / 180
	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(lat1*math.Pi/180)*math.Cos(lat2*math.Pi/180)*
			math.Sin(dLon/2)*math.Sin(dLon/2)
	return R * 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
}
